src/compiler/checker.ts

```diff
@@ -6835,6 +7028,23 @@ module ts {
                     });
                 }
             }
+            else if (node.kind === SyntaxKind.Decorator) {
+                let decorator = <Decorator>node;
+                switch (decorator.parent.kind) {
+                    case SyntaxKind.ClassDeclaration:
+                        args = [decorator.expression];
+                        break;
+                    case SyntaxKind.PropertyDeclaration:
+                        args = [decorator.expression, decorator.expression];
+                        break;
+                    case SyntaxKind.MethodDeclaration:
+                    case SyntaxKind.GetAccessor:
+                    case SyntaxKind.SetAccessor:
+                    case SyntaxKind.Parameter:
+                        args = [decorator.expression, decorator.expression, decorator.expression];

```

Comment:
I'm not crazy about this effective call arguments model. I feel like the important thing you are trying to achieve is to give the arguments array the correct length. But I'd almost rather you put undefined, rather than decorator.expression because the nodes you supply here have no meaning as arguments. The decorator.expression is really more like the call target than it is the argument.